Complete mapping of Prometheus alerts

aleksandra-galara opened the pull request:
It refers to #34 and complete mapping
alerts due to list created in issue

aleksandra-galara committed:
Complete mapping of Prometheus alerts

It refers to #34 and complete mapping
alerts due to list created in issue

Signed-off-by: Aleksandra Galara <a.galara@samsung.com>

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+    - name: KubeCronJobRunning
+      source_mapping:
+        origin: kubernetes
+        kind: cronjob
```
```suggestion
        kind: cron_job
```
Let's follow the convention of using underscores for kind names consisting of multiple parts. Note e.g. `persistent_volume_claim` of `daemon_set `.

bzurkowski replied:
```suggestion
        kind: cron_job
```
Let's follow the convention of using underscores for kind names consisting of multiple parts. Note e.g. `persistent_volume_claim` or `daemon_set `.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeCPUQuotaOvercommit
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        kind: namespace
+        properties:
+          name: namespace
+    - name: KubeVersionMismatch
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsDown
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: AlertmanagerDown
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: AlertmanagerDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: NodeExporterDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusOperatorDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
```
I'm a little bit hesitant about mapping these alerts to the cluster kind. They are not *very much* cluster-related, but rather Prometheus-related, and as such may disturb the field of view.

As far as I understand, alerts such as `PrometheusDown` or `NodeExporterDown` report that Prometheus instance was unable to reach scrape endpoints (unable to fetch metrics) - the problem occurred on the Prometheus-side.

Would it be possible to map the above alerts to the corresponding Prometheus instances (i.e. pods)?

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        properties:
+          name: pod
+          namespace: namespace
+    - name: PrometheusOperatorReconcileErrors
```
TODO: Recheck

bzurkowski replied:
I'm wondering what the `controller` label means in the [alert definition](https://github.com/coreos/kube-prometheus/blob/a8b4985de4dc8582dcaf8fd43d6f88a0c08c463d/manifests/prometheus-rules.yaml#L1761). Maybe together with the `namespace` label we could map it to something less generic than the namespace object?

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        kind: namespace
+        properties:
+          name: namespace
+    - name: PrometheusOperatorNodeLookupErrors
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        kind: namespace
+        properties:
+          name: namespace
+    - name: AlertmanagerConfigInconsistent
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        properties:
+          name: pod
+          namespace: namespace
+    - name: AlertmanagerMembersInconsistent
```
AlertmanagerMembersInconsistent

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: NodeNetworkInterfaceFlapping
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        kind: node
+        properties:
+          name: node
+    - name: KubeClientCertificateExpiration
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeDeploymentGenerationMismatch
```
Please, move it up near other deployment-related alerts.

bzurkowski replied:
Please, move it up near to other deployment-related alerts.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        properties:
+          name: deployment
+          namespace: namespace
+    - name: etcdInsufficientMembers
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: etcdGRPCRequestsSlow
```
TODO: Recheck

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdHighNumberOfFailedGRPCRequests
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdHighNumberOfFailedHTTPRequests
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdHighNumberOfFailedProposals
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdHighNumberOfLeaderChanges
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdMemberCommunicationSlow
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: etcdNoLeader
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
```
TODO: Recheck

bzurkowski replied:
I'm not sure whether `instance` label is actually what we think it is. It might be the IP/hostname of etcd pod rather than the name of a node. Look at this [issue](https://github.com/coreos/prometheus-operator/issues/1259) and [here](https://github.com/etcd-io/etcd/blob/322c38e16991754615e041f325d055cb9f673041/Documentation/etcd-mixin/test.yaml#L88) for examples.

If that's the case, do you think we could map these alerts by pod IP? We already [extract](https://github.com/openrca/orca/blob/a8e44c042aeef60bce3399a6baf4399b34b10139/orca/topology/infra/k8s/extractor.py#L57) an IP for each pod.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: persistent_volume_claim
+        properties:
+          name: persistentvolumeclaim
```
PVCs are namespaced. `namespace` property must be added to prevent conflicts.

bzurkowski replied:
PVCs are namespaced. In order to prevent conflicts you must include `namespace` property as well.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeAPIErrorBudgetBurn
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsListErrors
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsWatchErrors
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
```
Is it possible to map these alerts to corresponding pod instances?

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: AlertmanagerDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: NodeExporterDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusOperatorDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: TargetDown
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
+        properties:
+          name: namespace
```
I'm a little bit hesitant about mapping these alerts to the cluster kind. They are not *very much* cluster-related, but rather Prometheus-related, and as such may disturb the field of view.

As far as I understand, alerts such as `PrometheusDown` or `NodeExporterDown` report that Prometheus instance was unable to reach scrape endpoints (unable to fetch metrics) - the problem occurred on the Prometheus-side.

Would it be possible to map the above alerts to the corresponding Prometheus instances (i.e. pods)?

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeStateMetricsDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: AlertmanagerDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: NodeExporterDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: PrometheusOperatorDown
+      source_mapping:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
```
I'm a little bit hesitant about mapping these alerts to the cluster kind. They are not related to Kubernetes cluster, but rather Prometheus jobs/services, and as such may disturb the analytical field of view.

Alerts such as `PrometheusOperatorDown` or `NodeExporterDown` indicate that Prometheus instance was unable to scrape job endpoints (unable to fetch metrics). As such, similar alerts should be mapped to e.g. Prometheus service object that prevents metric scrape. Alternatively, we could map to Prometheus instances (pods) which reported job issues.

In the future we should think of a different mapping solution for this category of alerts. If you see other options that we could implement in this change, please comment. Otherwise, let's keep the mapping as it is and add a separate issue for mapping improvements.

aleksandra-galara replied:
Unfortunately, these are the only suggested changes that I couldn't make.
Alerts with "* Down" use absent() function, which doesn't return any labels ("returns an empty vector if the vector passed to it has any elements and a 1-element vector with the value 1 if the vector passed to it has no elements"), so I'll create a separate issue for mapping improvements.

bzurkowski replied:
I see. Then for now, let's keep the mapping as it is. Thanks for adding the issue for improvements.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+    - name: TargetDown
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
```
Is it possible to map this alert to service kind instead of namespace kind based on `service` + `namespace` [labels](https://github.com/coreos/kube-prometheus/blob/a8b4985de4dc8582dcaf8fd43d6f88a0c08c463d/manifests/prometheus-rules.yaml#L1728)?

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        properties:
+          name: pod
+          namespace: namespace
+    - name: NodeClockSkewDetected
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: NodeClockNotSynchronising
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
+    - name: NodeHighNumberConntrackEntriesUsed
+      source_mapping:
+        origin: kubernetes
+        kind: node
+        properties:
+          name: instance
```
Please, move this mapping up near to other [node-related alerts](https://github.com/openrca/orca/blob/master/helm/orca/config/alerts-mapping.yaml#L146-L187).

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        kind: persistent_volume_claim
+        properties:
+          name: persistentvolumeclaim
+    - name: AggregatedAPIDown
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
+        properties:
+          name: namespace
+    - name: AggregatedAPIErrors
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
+        properties:
+          name: namespace
```
For now, let's map aggregated API alerts to cluster kind.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeCPUQuotaOvercommit
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
+        properties:
+          name: namespace
+    - name: KubeMemoryQuotaOvercommit
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
+        properties:
+          name: namespace
```
Shouldn't we map these alerts to the cluster kind instead of the namespace kind?

Although the `kube_resourcequota` metric is namespace-scoped, [alert expressions](https://github.com/coreos/kube-prometheus/blob/a8b4985de4dc8582dcaf8fd43d6f88a0c08c463d/manifests/prometheus-rules.yaml#L1127-L1131) for these two alerts sum values for all metric instances (namespaces):
```yaml
- alert: KubeCPUQuotaOvercommit
    annotations:
    message: Cluster has overcommitted CPU resource requests for Namespaces.
    runbook_url: https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-kubecpuquotaovercommit
    expr: |
    sum(kube_resourcequota{job="kube-state-metrics", type="hard", resource="cpu"})
        /
    sum(kube_node_status_allocatable_cpu_cores)
        > 1.5
```
The alert name (`for Namespaces`) is misleading though..

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +185,416 @@ prometheus:
+    - name: PrometheusOperatorNodeLookupErrors
+      source_mapping:
+        origin: kubernetes
+        kind: namespace
```
Let's map to cluster kind instead of namespace kind.

Namespace kind is not enabled in the graph - we should avoid it as much as possible. Since many other Prometheus alerts (e.g. `PrometheusOperatorDown`) have been already mapped to cluster kind, we should be consistent.

bzurkowski reviewed (changes requested):
@aleksandra-galara First off, thanks for your first contribution. 🎉🍰  You did really well! 👌I added only a few minor remarks. Some are purely cosmetic, others concern the mapping scope.

I noticed, a significant number of alerts is mapped to the root cluster node. In most cases it's correct because these alerts don't provide enough labels to correlate with less generic elements. If you see any options to narrow down the mapping scope, please share your ideas 😉

aleksandra-galara committed:
Correct mapping of Prometheus alerts

It refers to #34 and complete mapping
alerts due to list created in issue.
It introduces the changes suggested
in the review

Signed-off-by: Aleksandra Galara <a.galara@samsung.com>

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +192,415 @@ prometheus:
+        origin: kubernetes
+        kind: cluster
+        properties: {}
+    - name: KubeMemoryQuotaOvercommit
```
This one is duplicated and should be removed.

bzurkowski commented on helm/orca/config/alerts-mapping.yaml:
```diff
@@ -185,6 +192,415 @@ prometheus:
+        properties:
+          name: service
+          namespace: namespace
+    - name: KubeClientErrors
```
I just looked at [this](https://github.com/coreos/prometheus-operator/issues/1631) Github thread and it seems that the `instance` label does not contain name of a node, but rather IP address of a client:
```
Kubernetes API server client 'kubelet/10.0.2.15:10250' is experiencing 3% errors.
Kubernetes API server client 'apiserver/192.168.99.100:8443' is experiencing 24% errors.
```
 I'm afraid we must map to the cluster node again.

bzurkowski reviewed (changes requested):
Thanks for the fixes. I added two more minor remarks. Thanks!

aleksandra-galara committed:
Correct mapping of Prometheus alerts(2)

It refers to #34 and complete mapping
alerts due to list created in issue.
It introduces the changes suggested
in the review

Signed-off-by: Aleksandra Galara <a.galara@samsung.com>

bzurkowski reviewed (approved):
Change approved. Ready for merge!
